src/mesh/voxel_cloud.rs

```diff
@@ -758,303 +491,316 @@ impl VoxelCloud {
+    block_start: &Point3<i32>,
+    voxel_dimensions: &Vector3<f32>,
+) -> Point3<f32> {
     assert!(
         !approx::relative_eq!(voxel_dimensions.x, 0.0)
             && !approx::relative_eq!(voxel_dimensions.y, 0.0)
             && !approx::relative_eq!(voxel_dimensions.z, 0.0),
         "Voxel dimensions can't be 0.0"
     );
     Point3::new(
-        (point.x / voxel_dimensions.x).round() as i32,
-        (point.y / voxel_dimensions.y).round() as i32,
-        (point.z / voxel_dimensions.z).round() as i32,
+        (relative_coordinate.x + block_start.x) as f32 * voxel_dimensions.x,
+        (relative_coordinate.y + block_start.y) as f32 * voxel_dimensions.y,
+        (relative_coordinate.z + block_start.z) as f32 * voxel_dimensions.z,
     )
 }
 
-/// Computes the voxel-space coordinates of a voxel containing the
-/// input point
-#[allow(dead_code)]
-fn cartesian_to_relative_voxel_coords(
+/// Computes the absolute voxel space coordinate of a voxel containing the input
+/// point.
+///
+/// # Panics
+/// Panics if any of the voxel dimensions is equal or below zero.
+fn cartesian_to_absolute_voxel_coordinate(
     point: &Point3<f32>,
-    block_start: &Point3<i32>,
-    voxel_dimensions: Vector3<f32>,
+    voxel_dimensions: &Vector3<f32>,
 ) -> Point3<i32> {
     assert!(
-        !approx::relative_eq!(voxel_dimensions.x, 0.0)
-            && !approx::relative_eq!(voxel_dimensions.y, 0.0)
-            && !approx::relative_eq!(voxel_dimensions.z, 0.0),
-        "Voxel dimensions can't be 0.0"
+        voxel_dimensions.x > 0.0 && voxel_dimensions.y > 0.0 && voxel_dimensions.z > 0.0,
+        "Voxel dimensions can't be below or equal to zero"
     );
     Point3::new(
-        (point.x / voxel_dimensions.x).round() as i32 - block_start.x,
-        (point.y / voxel_dimensions.y).round() as i32 - block_start.y,
-        (point.z / voxel_dimensions.z).round() as i32 - block_start.z,
+        (point.x / voxel_dimensions.x).round() as i32,
+        (point.y / voxel_dimensions.y).round() as i32,
+        (point.z / voxel_dimensions.z).round() as i32,
     )
 }
 
-/// Computes the center of a voxel in model-space coordinates from
-/// absolute voxel coordinates (relative to the voxel space origin).
-fn absolute_voxel_to_cartesian_coords(
-    absolute_coords: &Point3<i32>,
-    voxel_dimensions: Vector3<f32>,
-) -> Point3<f32> {
-    assert!(
-        !approx::relative_eq!(voxel_dimensions.x, 0.0)
-            && !approx::relative_eq!(voxel_dimensions.y, 0.0)
-            && !approx::relative_eq!(voxel_dimensions.z, 0.0),
-        "Voxel dimensions can't be 0.0"
-    );
-    Point3::new(
-        absolute_coords.x as f32 * voxel_dimensions.x,
-        absolute_coords.y as f32 * voxel_dimensions.y,
-        absolute_coords.z as f32 * voxel_dimensions.z,
-    )
+/// Computes an index to the linear representation of the voxel block from
+/// voxel coordinates relative to the voxel space block start.
+///
+/// Returns None if out of bounds.
```

Comment:
Now only the `to_one_dimensional_coordinate` fns return `None` if out of bounds. I wonder if this shouldn't be unified with the rest somehow.

Reply:
i find it correct. Only one when the dimensional coordinate is out of bounds, the related functions panic.

Reply:
Okay, it's internal functionality anyway